OpenRA.Mods.Common/Traits/World/HierarchicalPathFinder.cs

```diff
@@ -0,0 +1,759 @@
+
+				List<CPos> localPath;
+				using (var search = GetLocalPathSearch(
+					self, source, target, customCost, ignoreActor, check, laneBias, gridToSearch))
+					localPath = search.FindPath();
+
+				if (localPath.Count > 0)
+					return localPath;
+			}
+
+			RebuildDirtyGrids();
+
+			CPos? AbstractCellForLocalCell(CPos localCell) =>
+				gridInfos[GridIndex(localCell)].AbstractCellForLocalCell(localCell);
+
+			// We need to insert the source and target cells as nodes into the abstract graph.
+			// Determine costs to link these to the available abstract cells within their respective grids.
+			GraphEdge? EdgeFromLocalToAbstract(CPos localCell, CPos abstractCell)
+			{
+				// No edge needs to be inserted into the graph,
+				// the local cell matches an existing abstract cell.
+				if (localCell == abstractCell)
+					return null;
+
+				return new GraphEdge(localCell, abstractCell, costEstimator(localCell, abstractCell));
+			}
+
+			// The new edges will be treated as bi-directional.
+			var sourceAbstractCell = AbstractCellForLocalCell(source);
+			if (sourceAbstractCell == null)
+				return PathFinder.NoPath;
+			var targetAbstractCell = AbstractCellForLocalCell(target);
+			if (targetAbstractCell == null)
+				return PathFinder.NoPath;
+			var sourceEdge = EdgeFromLocalToAbstract(source, sourceAbstractCell.Value);
+			var targetEdge = EdgeFromLocalToAbstract(target, targetAbstractCell.Value);
+			var fullGraph = new AbstractGraphWithInsertedEdges(abstractGraph, sourceEdge, targetEdge);
+
+			// Determine an abstract path in both directions, for use in a bidirectional search.
+			var estimatedSearchSize = (abstractGraph.Count + 2) / 8;
+			using (var abstractSearch = PathSearch.ToTargetCellOverGraph(
+				fullGraph.GetConnections, locomotor, source, target, estimatedSearchSize))
+			{
+				if (!abstractSearch.ExpandToTarget())
+					return PathFinder.NoPath;
+
+				using (var reversedAbstractSearch = PathSearch.ToTargetCellOverGraph(
+					fullGraph.GetConnections, locomotor, target, source, estimatedSearchSize))
+				{
+					reversedAbstractSearch.ExpandToTarget();
+					return NavigatePathWithAbstractPathAsHeuristic(self, source, target, check, customCost, ignoreActor, laneBias, abstractSearch, reversedAbstractSearch);
+				}
+			}
+		}
+
+		/// <summary>
+		/// The abstract graph can become out of date when reachability costs for terrain change.
+		/// When this occurs, we must rebuild any affected parts of the abstract graph so it remains correct.
+		/// </summary>
+		void RebuildDirtyGrids()
+		{
+			if (dirtyGridIndexes.Count == 0)
+				return;
+
+			var customMovementLayers = world.GetCustomMovementLayers();
+			foreach (var gridIndex in dirtyGridIndexes)
+			{
+				var oldGrid = gridInfos[gridIndex];
+				var gridTopLeft = GetGridTopLeft(gridIndex, 0);
+				gridInfos[gridIndex] = BuildGrid(gridTopLeft.X, gridTopLeft.Y, customMovementLayers);
+				RebuildCostTable(gridTopLeft.X, gridTopLeft.Y, oldGrid, customMovementLayers);
```

Comment:
One thing I haven't quite got my head around: are we assuming / can we assume that the cost from grid A to B is the same as from B to A? If both A and B are dirty, is there a way to regenerate both and then calculate the costs only once?

Reply:
Yes the current cost estimator is assumed to return the same cost regardless of direction.

It feels like it should be possible to only rebuild the edges between dirty grids once. Will have a think and see if a clean solution is possible.

Reply:
This would be quite painful. The existing data structure and methods are centered around rebuilding a whole grid at once and swapping out a `List<GraphConnection>` wholesale.

We'd need some sort of different and clever data structure to allow us to quickly identify subsets of connections from one node that need to be added, updated or removed. At the same time the existing structure is pretty efficient for querying because we can do a single lookup in the dictionary and return the `List<GraphConnection>` that is consumed via `IPathGraph.GetConnections` - without any copies needed. Any new data structure would want to keep this efficiency as much as possible (or possibly rework that interface to preserve it).

Making this change would improve update performance, but ensuring querying performance gets no worse it a much more important property to preserve, and I am struggling to think of a method that would allow us to meet both those goals. On that basis I will leave this as-is for now.